common/certificate/src/validate_device_id.rs

```diff
@@ -0,0 +1,75 @@
+use crate::CertificateError;
+
+fn is_valid_device_id_char(ch: char) -> bool {
+    matches!(ch, 'A'..='Z'
+    | 'a'..='z'
+    | '0'..='9'
+    | '\''
+    | '='
+    | '('
+    | ')'
+    | '+'
+    | ','
+    | '-'
+    | '.'
+    | ':'
+    | '?')
+}
+
+pub fn is_valid_device_id(id: &str, max_cn_size: usize) -> Result<(), CertificateError> {
+    if id.is_empty() {
+        return Err(CertificateError::EmptyName);
+    } else if id.len() > max_cn_size {
+        return Err(CertificateError::TooLongName {
+            name: id.into(),
+            max_cn_size,
+        });
+    } else if !id.chars().all(is_valid_device_id_char) || id.contains(char::is_control) {
+        return Err(CertificateError::InvalidCharacter { name: id.into() });
+    }
+
+    Ok(())
+}
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn valid_device_id_with_aplhabets() {
+        assert!("ThinEdgeDevice".chars().all(is_valid_device_id_char));
+    }
+
+    #[test]
+    fn valid_device_id_with_digits() {
+        assert!("123ThinEdgeDevice".chars().all(is_valid_device_id_char));
+    }
+    #[test]
+    fn valid_device_id_with_special_chars() {
+        assert!("'?:=()+,-.:123ThinEdgeDevice-id"
+            .chars()
+            .all(is_valid_device_id_char));
+    }
+    #[test]
+    fn invalid_device_ids() {
+        for id in vec![
+            "thinedge>",
+            "thinedge<",
+            "thinedge@",
+            "thinedge#",
+            "thinedge|",
+            "thinedge[",
+            "thinedge]",
+            "thinedge}",
+            "thinedge{",
+            "thinedge^",
+            "thinedge*",
+            "thinedge&",
+            "thin_edge",
```

Comment:
Oh! This "_" is invalid? This is not handy! 

* Same things with the `@` that prevents using e-mail like identifiers.
* By contrast, it's strange to accept parenthesis in a name. Why "foo(bar)" would be okay and not "foo{bar}" ?
* This needs to be discussed with @itsyitsy

Reply:
Add `thin/edge` also to this set as excluding this character is critical since MQTT is an integral part of our design and we wouldn't want the MQTT topic separator in a name. `$` also might be a good candidate to exclude.

Reply:
Yes, X509 does not allow it, as per this link https://datatracker.ietf.org/doc/html/rfc5280, I feel we can use it. But not sure.
A para from the above link. I feel need more research on this.
"Implementers should note that the at sign ('@') and underscore ('_')    characters are not supported by the ASN.1 type PrintableString.   These characters often appear in Internet addresses.  Such addresses   MUST be encoded using an ASN.1 type that supports them.  They are usually encoded as IA5String in either the emailAddress attribute within a distinguished name or the rfc822Name field of GeneralName.   Conforming implementations MUST NOT encode strings that include   either the at the sign or underscore character as PrintableString."

Reply:
Now I got some answer for my question from OpenSSL. So, x509 will allow all the UTF8String.

X520CommonName ::= CHOICE {
      teletexString     TeletexString   (SIZE (1..ub-common-name)),
      printableString   PrintableString (SIZE (1..ub-common-name)),
      universalString   UniversalString (SIZE (1..ub-common-name)),
      utf8String        UTF8String      (SIZE (1..ub-common-name)),
      bmpString         BMPString       (SIZE (1..ub-common-name)) }